crates/miden-objects/src/transaction/inputs.rs

```diff
@@ -531,3 +535,170 @@ pub fn validate_account_seed(
+            account_witness: merkle_proof,
+            storage_map_proofs,
+        }
+    }
+
+    /// Returns the account's [`AccountId`]
+    pub fn id(&self) -> AccountId {
+        self.account_header.id()
+    }
+
+    /// Returns the account's [`AccountHeader`]
+    pub fn account_header(&self) -> &AccountHeader {
+        &self.account_header
+    }
+
+    /// Returns the account's [`AccountStorageHeader`].
+    pub fn storage_header(&self) -> &AccountStorageHeader {
+        &self.storage_header
+    }
+
+    /// Returns the account's storage maps.
+    pub fn storage_map_proofs(&self) -> &[SmtProof] {
+        &self.storage_map_proofs
+    }
+
+    /// Returns the account's [`AccountCode`].
+    pub fn account_code(&self) -> &AccountCode {
+        &self.account_code
+    }
+
+    /// Returns the account witness.
+    pub fn account_witness(&self) -> &MerklePath {
+        &self.account_witness
+    }
+
+    /// Computes account root based on the account witness.
+    pub fn compute_account_root(&self) -> Result<Digest, MerkleError> {
+        self.account_witness()
+            .compute_root(self.id().prefix().into(), self.account_header().commitment())
+    }
+
+    /// Extends the storage proofs with the input `smt_proofs` and returns the new structure
+    #[must_use]
+    pub fn with_storage_map_proofs(
+        mut self,
+        smt_proofs: impl IntoIterator<Item = SmtProof>,
+    ) -> Self {
+        self.storage_map_proofs.extend(smt_proofs);
+        self
+    }
+
+    /// Consumes the [`ForeignAccountInputs`] and returns its parts.
+    pub fn into_parts(
+        self,
+    ) -> (AccountHeader, AccountStorageHeader, AccountCode, MerklePath, Vec<SmtProof>) {
+        (
+            self.account_header,
+            self.storage_header,
+            self.account_code,
+            self.account_witness,
+            self.storage_map_proofs,
+        )
+    }
+}
+
+impl Serializable for ForeignAccountInputs {
+    fn write_into<W: crate::utils::ByteWriter>(&self, target: &mut W) {
+        self.account_header.write_into(target);
+        self.storage_header.write_into(target);
+        self.account_code.write_into(target);
+        self.account_witness.write_into(target);
+        self.storage_map_proofs.write_into(target);
+    }
+}
+
+impl Deserializable for ForeignAccountInputs {
+    fn read_from<R: crate::utils::ByteReader>(
+        source: &mut R,
+    ) -> Result<Self, crate::utils::DeserializationError> {
+        let account_header = AccountHeader::read_from(source)?;
+        let storage_header = AccountStorageHeader::read_from(source)?;
+        let account_code = AccountCode::read_from(source)?;
+        let merkle_proof = MerklePath::read_from(source)?;
+        let storage_maps = Vec::<SmtProof>::read_from(source)?;
+        Ok(ForeignAccountInputs::new(
+            account_header,
+            storage_header,
+            account_code,
+            merkle_proof,
+            storage_maps,
+        ))
```

Comment:
Should we do some validation here? For example?

- `account_header.storage_commitment() == storage_header.commitment()`
- `account_header.code_commitment() == account_code.commitment()`
- Maybe something else as well? 

On the other hand, these checks may be a bit expensive - so, maybe better add something like `AccountInputs::validate()` method?

Reply:
I added an `AccountInputs ::is_valid(&self) -> bool`. If you think it should be more granular I can change it to `validate()` and have it return an error type. For this I also had to add a way to compute the commitment in `AccountStorageHeader`.